HW-2 Model 적용

chuuuul opened the pull request:
HW-2 Model 적용

 - Control은 Model에서 데이터를 가져온다.
 - Model은 Repository를 통해서 Local Data와 Remote Data를 가져온다.
 - Local Data Source는 Shared Preferences로 데이터를 가져온다.
 - Remote Data Source는 Retrofit을 통해서 데이터를 가져온다.
 - Repository는 싱글턴으로 관리한다.

chuuuul committed:
HW-2 리펙토링: 패키지 폴더 수정

 - Model의 구조 변경

chuuuul committed:
HW-2 리팩토링 : 파일명 변경

 - MovieVO -> MovieModel
 - MovieResponseVO -> MovieResponseModel

chuuuul committed:
HW-2 리팩토링 : Remote Data Source 인터페이스 작성

chuuuul committed:
HW-2 리팩토링 : 컬렉션 타입 변경

 - ArrayList -> MutableList 로 변경

chuuuul committed:
HW-2 리팩토링 : Model의 구조 변경(Remote)

 - 컨트롤에서는 Model에서 데이터를 가져온다.
 - 해당 모델은 Repository로써 Repository에서는 Local과 Remote를 관리한다.
 - Repository는 싱글턴으로 관리하여, 다른곳에서도 같은 데이터를 접근 할 수 있다.

chuuuul committed:
HW-2 리팩토링 : Util 폴더 생성

ConstValue 파일 이동
Remote Data Source 인터페이스 작성

chuuuul committed:
HW-2 기능 : Const 변수 추가

- val -> const val
- Movie Shared Name 추가

chuuuul committed:
HW-2 기능 : Naver Local Data Source 작성

 - Local Data를 Shared Preferences로 가져오기 및 저장

chuuuul committed:
HW-2 형식 : 실패 로그 변경

developer-sungjae commented:
안녕하세요 늦게 코멘트 드려 죄송합니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/local/NaverLocalDataSourceImpl.kt:
```diff
@@ -0,0 +1,66 @@
+        keyword: String,
+        onSuccess: (movies: List<MovieModel>) -> Unit
+    ) {
+        val sharedPref =
```
각 메서드마다 SharedPreference 객체를 각각 계속 생성해주셨는데 이렇게 하신 이유가 context를 '메서드 호출시에만 부를 수 있어서'라면 NaverLocalDataSourceImpl (context: Context)형태로 받아서 property로 선언하면 중복코드를 없앨 수 있을 것 같은데 어떻게 생각하시나요?
제가 예측한것이 아니라면 혹시 각 객체를 따로따로 생성하시는 이유가 무엇인지 궁금합니다.

두번째로는 각 객체 생성시  ?: return으로 null처리를 해주셨습니다만 null인 경우는 어떠한 경우인건지 궁금합니다.

chuuuul replied:
> 각 메서드마다 SharedPreference 객체를 각각 계속 생성해주셨는데 이렇게 하신 이유가 context를 '메서드 호출시에만 부를 수 있어서'라면 NaverLocalDataSourceImpl (context: Context)형태로 받아서 property로 선언하면 중복코드를 없앨 수 있을 것 같은데 어떻게 생각하시나요?

7b6ed8a, fbd334c
감사합니다. 수정했습니다.
context를 계속 받은 이유는 원래 Repository를 object로 생성하여 싱글턴으로 운용하려고 했습니다. Object 클래스로 만들어서 remote와 local의 data source를 가지고 있으면서 액티비티에 종속되지 않은 Model 객체를 생성하려고 했습니다. 그래서 액티비티에서 context를 받지 않고 필요할때마다 context를 매번 불러주어 사용하러고 했습니다.
 말씀해주신 지적사항을 적용하려고 하니깐 Object로 선언한 클래스에는 생성자를 받을 수 없어 context를 local 클래스에 생성자로 넣어 줄 수가 없었습니다. 그래서 액티비티에 종속되는 Class를 사용하기로 하고 말씀해주신 대로 context를 프로퍼티로 사용하였습니다. Model을 Activity마다 갖는 경우 말고 싱글턴으로 가져가는 경우는 없는지 궁금하네요 ㅎㅎ

> 두번째로는 각 객체 생성시 ?: return으로 null처리를 해주셨습니다만 null인 경우는 어떠한 경우인건지 궁금합니다.

저도 자세하게 모르겠지만
`context.getSharedPreferences(MOVIE_SHARED_PREFERENCE_NAME, Context.MODE_PRIVATE)`
의 반환값은 SharedPreferences! 가 됩니다. 즉 null 처리가 필요하다고 판단되었고, 공식문서에서도 null처리가 포함된 예제가 있었습니다.
 추측하건데, Context가 유효하지 않을 경우 null이 발생하지 않을까 싶습니다.

developer-sungjae replied:
> 
말씀해주신 지적사항을 적용하려고 하니깐 Object로 선언한 클래스에는 생성자를 받을 수 없어 context를 local 클래스에 생성자로 넣어 줄 수가 없었습니다. 그래서 액티비티에 종속되는 Class를 사용하기로 하고 말씀해주신 대로 context를 프로퍼티로 사용하였습니다. Model을 Activity마다 갖는 경우 말고 싱글턴으로 가져가는 경우는 없는지 궁금하네요 ㅎㅎ
>

SageHabitus replied:
> 말씀해주신 지적사항을 적용하려고 하니깐 Object로 선언한 클래스에는 생성자를 받을 수 없어 context를 local 클래스에 생성자로 넣어 줄 수가 없었습니다. 그래서 액티비티에 종속되는 Class를 사용하기로 하고 말씀해주신 대로 context를 프로퍼티로 사용하였습니다. Model을 Activity마다 갖는 경우 말고 싱글턴으로 가져가는 경우는 없는지 궁금하네요 ㅎㅎ

작성하신 의도는 알고 있었습니다.
하지만 아쉬운 중복코드가 발생해서 코멘트 드렸습니다.
저같은 경우는 예전에 종속성을 조금이라도 더 떼어놓으려고 하다보니 생성자로 초기화 하지 않고 Object의 Property로 Context를 초기화하였습니다.
그러면 메소드 호출시에 context를 매번 parameter로 안넣어도 되며 최초 생성시에만 context를 세팅하면 가능하다고 보고 있습니다.

1) 위에 언급한대로 원래 의도대로 하셔서 보완해주시거나 아니면 
2) 제가 작성한 커스텀 클래스 이용해서 보완해주시면 좋을 것 같습니다.

(사실 어느 방법을 하더라도 사실 지금 단계는 완벽하게 떼어놓을 수 없을 것 같습니다.)

`
object SharedPreferenceFactory {

    private lateinit var sharedPref: SharedPreferences

    fun create(context: Context) {
        sharedPref = context.getSharedPreferences("Name", MODE_PRIVATE)
    }

    fun getString(KEY: String, defaultValue: String?): String? {

        return sharedPref.getString(KEY, defaultValue)
    }

    fun getInt(KEY: String, defaultValue: Int): Int {

        return sharedPref.getInt(KEY, defaultValue)
    }

    fun getBoolean(KEY: String, defaultValue: Boolean): Boolean {

        return sharedPref.getBoolean(KEY, defaultValue)
    }

    fun putString(KEY: String, value: String?) {
        sharedPref.edit().apply {
            putString(KEY, value)
            apply()
        }
    }

    fun putInt(KEY: String, value: Int) {
        sharedPref.edit().apply {
            putInt(KEY, value)
            apply()
        }
    }

    fun putBoolean(KEY: String, value: Boolean) {
        sharedPref.edit().apply {
            putBoolean(KEY, value)
            apply()
        }
    }

    fun clear() {
        sharedPref.edit().apply {
            clear()
            apply()
        }
    }
}
`

chuuuul replied:
c98e4c1
감사합니다. companion object를 써서 한번 구현해봤는데 어떤가요?

SageHabitus replied:
[코틀린 컨벤션 참고 - 이건 저도 헷갈려서 이참에 정리하면서 공유드립니다.]
레이아웃
일반적으로 클래스의 내용은 다음 순서로 정렬됩니다.

Property declarations and initializer blocks
Secondary constructors
Method declarations
Companion object
Do not sort the method declarations alphabetically or by visibility, and do not separate regular methods from extension methods. Instead, put related stuff together, so that someone reading the class from top to bottom can follow the logic of what's happening. Choose an order (either higher-level stuff first, or vice versa) and stick to it.


쉽게 말하면 프로퍼티, init블록
그다음 부생성자
메서드 선언나열
컴패니언 옵젝트 인것 같습니다.
코틀린 컨벤션에 따라 바꿔주시면 좋을 것 같습니다.

chuuuul replied:
d064ad8, c7ae081
감사합니다. 수정했습니다. 또 하나 배워갑니다!

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/local/NaverLocalDataSourceImpl.kt:
```diff
@@ -0,0 +1,66 @@
+        val movieString = movieModelToString(movies)
+
+        sharedPref.edit().apply {
+            putString(keyword, movieString)
```
개인적인 의견입니다. 협업의 관점에서 보았을 때,
movieString는 변수명을 보면 movieString이 뭐지? 왔다갔다 할 것 같다는 생각이 듭니다.
한번의 딜레이가 있을 것 같습니다만, 더 직관적으로 parce관련된 용어를 섞어서 쓰시면 좋지 않을까 합니다.
생각해보시고 아니라고 생각하시면 안바꾸셔도 됩니다.

또한 movieModelToString,stringToMovieModel에서 parameter는 List형태인데 Model보다는 List 또는 다른것을 쓰는게 더 직관적이고 좋지 않을까 합니다.

chuuuul replied:
10560f9
감사합니다. 수정했습니다.
parce라는 단어가 다가오지가 않아서 jsonMovieData로 변경하였습니다!

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/local/NaverLocalDataSourceImpl.kt:
```diff
@@ -0,0 +1,66 @@
+        return Gson().toJson(movies)
+    }
+
+    private fun stringToMovieModel(gsonString: String): List<MovieModel> {
```
json데이터로부터 Gson을 이용하여 해당 객체에 맞게 셋팅한다 라는 의미를 내포하고 있습니다만 gsonString보다는 jsonString 또는 jsonData, jsonValue, parcedJsonValue 등 다른 것을 쓰는게 좋을 것 같습니다.

chuuuul replied:
f35b1b3
감사합니다. 변수명을 저도 헷갈려서 잘못지었네요. 수정했습니다!

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/local/NaverLocalDataSourceImpl.kt:
```diff
@@ -0,0 +1,66 @@
+            context.getSharedPreferences(MOVIE_SHARED_PREFERENCE_NAME, Context.MODE_PRIVATE)
+                ?: return
+
+        val movieString = sharedPref.getString(keyword, null) ?: return
```
궁금해서 질문드립니다.
만약에 keyword의 데이터가 아무것도 없다면 null이 movieString에 저장 되나요? 아니면 null 때문에 return이 되는것인가요? 혹시 keyword값이 단순 "" 같은 빈문자열일 경우는 없나요?

chuuuul replied:
> keyword의 데이터가 아무것도 없다면 null이 movieString에 저장 되나요? 
> 아니면 null 때문에 return이 되는것인가요?

 Shared Preference는 key : value 형식으로 받아오는데 key에 해당되는 데이터가없다면 2번째 인자로 들어간 Default Value가 튀어나오게 됩니다. 
 해당코드는 Default Value를 null로 지정하여 Key에 맞는 데이터가 없다면 null이 나오게 되고 종료 분기를 타도록 하였습니다. Default Value가 ""일 필요가 없다고 판단하였습니다.

> keyword값이 단순 "" 같은 빈문자열일 경우는 없나요?

검색할때 공백문자는 들어가지 않도록 검증 코드가 들어가 있습니다.
`if (isValidKeyword(keyWord)) { ... }`

developer-sungjae commented:
개인적인 의견입니다.
전체적으로 context를 모든 메서드마다 받으셨는데 property로 선언하고 paramter를 지우면 더 깔끔하지 않을까 하는데 어떠신가요?

developer-sungjae commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/repository/NaverRepository.kt:
```diff
@@ -0,0 +1,16 @@
+interface NaverRepository {
+
+    fun getMovieList(
+        context: Context,
```
context property

developer-sungjae commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/repository/NaverRepositoryImpl.kt:
```diff
@@ -0,0 +1,42 @@
+    private val naverLocalDataSourceImpl = NaverLocalDataSourceImpl()
+
+    override fun getMovieList(
+        context: Context,
```
context property

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/local/NaverLocalDataSource.kt:
```diff
@@ -0,0 +1,20 @@
+
+interface NaverLocalDataSource {
+    fun getMovieList(
+        context: Context,
```
context property

chuuuul replied:
7b6ed8a, fbd334c
context 관련 수정했습니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/MovieListAdapter.kt:
```diff
@@ -50,7 +50,7 @@ class MovieListAdapter :
     }
 
-    fun addNewItems(movieList : ArrayList<MovieVO>){
+    fun addNewItems(movieList: ArrayList<MovieModel>) {
```
movies인 부분이 있고 moviesList인 부분이 있습니다.
각각은 List형태로 Type이 비슷한데 헷갈릴 염려가 있을 것 같습니다. 통일 부탁드립니다.

chuuuul replied:
6c75701
감사합니다. 수정했습니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/repository/NaverRepositoryImpl.kt:
```diff
@@ -0,0 +1,42 @@
+        onSuccess: (movies: List<MovieModel>) -> Unit,
+        onFailure: (t: Throwable) -> Unit
+    ) {
+        naverLocalDataSourceImpl.getMovieList(context, keyword, onSuccess)
```
로컬과 리모트 데이터 가져오는 메서드 사이에 한칸 띄어주세요.

chuuuul replied:
23addea
감사합니다. 수정했습니다.

developer-sungjae commented:
commit에는 없으나 adapter 보고 코멘트 남깁니다.
65line]  val url = if (imageUrl.isBlank()) NO_IMAGE_URL else imageUrl

72line] private fun removeMarkupTag(string: String): String {
        return string.replace("<b>", "").replace("</b>", "")
    }

65번에서는 if, else brace 처리 부탁드립니다. 가독성에 좋지 않습니다.

72번에서는 removeMarkupTag 메서드 자체가 한군데에서만 쓰이는 메서드이고 저 <b>태그만으로는 못푸는 Html들이 있어서 혹시 이게 재활용성이 있는지 궁금합니다.
또한
removeMarkupTag(movie.title) 로 호출하셨는데 input parameter는 title인데
메서드 parameter는 string : String이라서 title만을 파싱하는 메서드의 parameter 이름으로 생각해보았을 때 title과 바로 매치가 안될뿐더러 string :String 이란 파라미터는 별로 의미없는 이름인 것 같습니다. 수정 부탁드립니다.

developer-sungjae commented:
그럼 고생하세요.

chuuuul committed:
HW-2 형식 : 메소드간 줄바꿈 추가

chuuuul committed:
HW-2 리팩토링 : 파라미터 이름 규칙 변경 ( movies -> movieList )

chuuuul committed:
HW-2 리팩토링 : Json -> Gson 메서드 AppUtill로 이동

 - 제네릭 사용하여 공용 코드로 변경
 - stringToMovieModel -> convertJsonToGson 메서드명 변경
 - movieModelToString -> convertGsonToJson 메서드명 변경

chuuuul committed:
HW-2 리팩토링 : 변수명 변경 ( movieString -> jsonMovieData )

chuuuul committed:
HW-2 리팩토링 : context에 관련된 공통 코드 제거

 - NaverLocationDataSource에서 context를 생성자로 받는다.
 - 생성자로 받은 context로 Shared Preferences 객체를 가져온다.

chuuuul committed:
HW-2 버그 : convertJsonToGson 메서드

 - com.google.gson.internal.LinkedTreeMap 관련 에러 발생
 - 제네릭으로 json에서 gson으로 변환하는 과정중에 에러 발생
 - 파라미터로 클래스 이름을 받아야 된다.

chuuuul committed:
HW-2 리팩토링 : NaverRepository를 Object에서 클래스로 변경

 - Repository를 싱글턴에서 클래스로 변경
 - Local Data Source에서 Context를 프로퍼티로 관리
 - Context에 관한 중복코드 제거

chuuuul committed:
HW-2 형식 : Bracket으로 가독성 개선

chuuuul committed:
HW-2 형식 : 파라미터 명명 변경

 - removeMarkupTag(string : String) -> removeMarkupTag(html : String)

chuuuul committed:
HW-2 형식 : 파라미터 명명 변경

 - removeMarkupTag(string : String) -> removeMarkupTag(html : String)

chuuuul commented:
File Changed 에서 MovieListAdapter.kt파일 아래에 Expend All 아이콘이 있습니다. 그걸로 해당 코드가 보이네요

> [65line]
> val url = if (imageUrl.isBlank()) NO_IMAGE_URL else imageUrl
> 65번에서는 if, else brace 처리 부탁드립니다. 가독성에 좋지 않을 것 같습니다.

c66b034
수정했습니다.

> 
> [72line]
> private fun removeMarkupTag(string: String): String {
> return string.replace("<b>", "").replace("</b>", "")
> }


> 72번에서는 removeMarkupTag 메서드 자체가 한군데에서만 쓰이는 메서드이고 저 bold 태그만으로는 못푸는 Html들이 있어서 혹시 이게 재활용성이 있는지 궁금합니다.
> 또한
> removeMarkupTag(movie.title) 로 호출하셨는데 input parameter는 title인데
> 메서드 parameter는 string : String이라서 title만을 파싱하는 메서드의 parameter 이름으로 생각해보았을 때 title과 바로 매치가 안될뿐더러 string :String 이란 파라미터는 별로 의미없는 이름인 것 같습니다. 수정 부탁드립니다.

652261b
수정했습니다.

네이버 영화 검색 API 문서를 보면 다음과 적혀있습니다.

_검색 결과 영화의 제목을 나타낸다. 제목에서 검색어와 일치하는 부분은 태그로 감싸져 있다._

해당 내용을 Bold 태그가 들어간다는 것으로 이해했고, 그 외의 HTML 코드는 처리하지 않도록 했습니다. 재활용성이 없다면 inline 처리를 하라는 말씀이신지 모르겠지만 inline을 쓰려니깐 효과가 미미하다고 떠서 추가하진 않겠습니다.

 파라미터 이름은 지금은 title에서만 사용하지만 일반적으로 string에서 html 코드를 지우는 작업이기 때문에 string이라고 변수명을 지었는데, 굳이 title이라는 파라미터 명으로 사용할 필요성을 못느껴서 변수명만 성의 있게 변경하겠습니다.

developer-sungjae commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/util/AppUtil.kt:
```diff
@@ -0,0 +1,15 @@
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+
+object AppUtil {
```
kotlin에서는 함수만 선언 할 수 있기 때문에 굳이 object로 안감싸도 될 것 같습니다.
함수만 선언하면 정적으로 선언됩니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/repository/NaverRepositoryImpl.kt:
```diff
@@ -0,0 +1,41 @@
+import com.example.architecture.data.source.local.NaverLocalDataSourceImpl
+import com.example.architecture.data.source.remote.NaverRemoteDataSourceImpl
+
+class NaverRepositoryImpl(val context: Context) : NaverRepository {
```
private 추가해주세요

chuuuul replied:
dc78dd8
감사합니다. 수정했습니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/util/AppUtil.kt:
```diff
@@ -0,0 +1,15 @@
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+
+object AppUtil {
```
아시겠지만 참고로 말씀드리면 object를 안쓰더라도 사용이 가능한데 object로 선언한 이유가 무엇인지 여쭤봐도 될까요?

chuuuul replied:
제 개인적인 취향인데, Const 변수를 관리하는 것과 비슷하다고 생각합니다. 앱 전반에 사용 하게 될 수 있는 코드를 AppUtil에 넣어서 사용합니다. 그러다보니 사용할때마다 AppUtil 객체를 만들어서 사용하는것보다 Object로 생성해서 사용하는 편이 더 효율적이라고 생각했습니다.

SageHabitus replied:
정답은 없고 궁금해서 여쭤봤습니다. 감사합니다. :)

SageHabitus replied:
아실수도 있겠지만 혹시나 메서드만 선언하는 것을 모르실까 하여 여쭤봤습니다 :) 
철님의 논리가 있으시니 그게 맞는 것 같습니다 :)

developer-sungjae commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/remote/NaverRemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,50 @@
+                call: Call<MovieResponseModel>,
+                response: Response<MovieResponseModel>
+            ) {
+                val movieResponse = response.body()
+
+                movieResponse?.also {
+                    onSuccess(it.movieList)
+                    saveMovieInLocal(keyWord, it.movieList)
+                }
```
1)
```suggestion
                movieResponse?.movieList?.let {
                    onSuccess(it)
                    saveMovieInLocal(keyWord, it)
                }
```
2)
```suggestion
                movieResponse?.movieList?.also {
                    onSuccess(it)
                    saveMovieInLocal(keyWord, it)
                }
```

SageHabitus replied:
1)
```suggestion
                movieResponse?.movieList?.let {
                    onSuccess(it)
                    saveMovieInLocal(keyWord, it)
                }
```
2)
```suggestion
                movieResponse?.movieList?.also {
                    onSuccess(it)
                    saveMovieInLocal(keyWord, it)
                }
```
3)
```suggestion
                val movieResponse = response.body()?.movieList

                movieResponse?.also {
                    onSuccess(it)
                    saveMovieInLocal(keyWord, it)
                }
```

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/remote/NaverRemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,48 @@
+
+class NaverRemoteDataSourceImpl : NaverRemoteDataSource {
+
+    private val retrofit by lazy {
```
NaverRemoteDataSoruceImpl에 property 선언 후 override로 구현 해주세요.

chuuuul replied:
211e96e
감사합니다. 수정했습니다.

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/remote/NaverRemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,48 @@
+
+class NaverRemoteDataSourceImpl : NaverRemoteDataSource {
+
+    private val retrofit by lazy {
+        Retrofit.Builder()
+            .baseUrl(ConstValue.MOVIE_SEARCH_API_URL)
+            .addConverterFactory(GsonConverterFactory.create())
+            .build()
+    }
```
```suggestion
    override val retrofit: Retrofit by lazy {
        Retrofit.Builder()
            .baseUrl(ConstValue.MOVIE_SEARCH_API_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
    }
```

SageHabitus commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/local/NaverLocalDataSourceImpl.kt:
```diff
@@ -0,0 +1,50 @@
+
+class NaverLocalDataSourceImpl(private val context: Context) : NaverLocalDataSource {
+
+    private val sharedPref: SharedPreferences by lazy {
```
by lazy로 한 이유가 있을까요?

chuuuul replied:
개인적인 취향인데 사용할때 초기화 하려고 by lazy로 선언했습니다. 액티비티에 들어갈때 생성 하기보다는 검색할때 생성하고 싶었습니다.

SageHabitus replied:
정답은 없고 궁금해서 여쭤봤습니다. 감사합니다.

chuuuul committed:
HW-2 리팩토링 : 가독성 개선

chuuuul committed:
Merge remote-tracking branch 'origin/2005/chuuuul-HW-2' into 2005/chuuuul-HW-2

chuuuul committed:
HW-2 리팩토링 : 생성자 context를 private로 관리

chuuuul committed:
HW-2 리팩토링 : Retrofit 객체 interface에 추가

chuuuul committed:
HW-2 리팩토링 : Model을 Companion Object로 관리

chuuuul committed:
HW-2 리팩토링 : companion object 내 객체 private로 수정

 - 싱글턴으로 구현했지만 직접 접근할 필요가 없기 때문에 private로 수정

chuuuul committed:
HW-2 형식 : Kotlin Convention 적용

  - Kotlin Convention에 따라 companion object 위치 수정

developer-sungjae commented:
고생하셨습니다.

sjjeong commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/SearchActivity.kt:
```diff
@@ -5,29 +5,17 @@ import android.util.Log
-            .addConverterFactory(GsonConverterFactory.create())
-            .build()
-    }
+    private val movieList = mutableListOf<MovieModel>()
```
이거는 불필요해보이네요

chuuuul replied:
68e5bda
감사합니다. 수정했습니다.

sjjeong commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/repository/NaverRepositoryImpl.kt:
```diff
@@ -0,0 +1,48 @@
+            keyword,
+            onSuccess,
+            onFailure,
+            this::saveMovieList
```
이 작업을 RemoteDataSource에 보내서 하는게 아니라 Repository에서 하도록 해주세요
getMovieList(keyword, {}, onFailure) 이렇게 하면 됩니다.

chuuuul replied:
ed5a7aa
감사합니다. 수정했습니다!

sjjeong commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/local/NaverLocalDataSourceImpl.kt:
```diff
@@ -0,0 +1,50 @@
+
+        val jsonMovieData = convertGsonToJson(movieList)
+
+        sharedPref.edit().apply {
+            putString(keyword, jsonMovieData)
+            apply()
+        }
```
apply는 불필요합니다
```kotlin
sharedPref.edit()
    .putString(keyword, jsonMovieData)
    .apply()
```
아래 clearData도 수정해주세요

chuuuul replied:
ae9a6c1
감사합니다. 수정했습니다.

sjjeong commented on 2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/util/AppUtil.kt:
```diff
@@ -0,0 +1,15 @@
+        return Gson().toJson(gsonData)
+    }
+
+    fun <T> convertJsonToGson(jsonData: String, clazz: Class<T>): Collection<T> {
```
reified에 대해 알아보시고 clazz 파라미터를 제거해주세요

chuuuul replied:
2c5ccb5
감사합니다. 수정했습니다!
예전에 시도했었는데 다른부분에서 오류났던것을 안되는걸로 착각했었네요..

chuuuul committed:
HW-2 리팩토링 : 불필요한 변수 제거

 - activity의 muvieList 제거

chuuuul committed:
HW-2 리팩토링 : apply관련 가독성 개선

 - SharedPreferences의 apply와 Scope의 apply가 혼동 되어 scope의 apply 제거

chuuuul committed:
HW-2 리팩토링 : convertJsonToGson의 파라미터 수정

 - class 타입을 파라미터로 받던것을 reified로 대체하여 파라미터 개수를 줄임
 - reified는 타입 제네릭을 런타임에도 접근 할 수 있도록 도와준다.

chuuuul committed:
HW-2 리팩토링 : RemoteDataSource의 Local Save 부분 제거

 - RemoteDataSource의 on Success에 해당 작업을 넣음

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
